### PR TITLE
[Booleans] new concept

### PR DESCRIPTION
**Pull request by colinleach**
For discussion on the [forum](https://forum.exercism.org/t/julia-learning-track-redux/12826/16).

This concept draft borrows heavily from `exercises/concept.wip/.docs/introduction.md`, with some edits and new additions. It pairs with `annalyns-infiltration`.

### CODE REVIEW

File: concepts/booleans/about.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+If `x` was 3 they would both be `true`.
+We can express statements like "is x less than 5 and not equal to y?" using [Boolean operators][boolean-operators]: `!` (not), `&&` (and), `||` (or).
+
+In Julia (and many other programming languages), `&&` has a [higher precedence][operator-precedence] than `||` (in the same way that `*` is applied before `+`).
+This means that `true || false && true` evaluates to `true` because it is parsed as `(true || false) && true`.
+It is common to include explicit brackets anyway so that the reader doesn't need to think about this.
+
+### Logical _not_
+
+`!` represents the logical "not" operation in  Julia.
```
Comment (BNAndras) [inline concepts/booleans/about.md:31]:
```suggestion
`!` represents the logical "not" operation in Julia.
```

File: concepts/booleans/about.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+## Boolean logic
+
+Imagine we have the following Boolean expressions in Julia: `5 > x` and `x != 0`.
+If `x` was 3 they would both be `true`.
+We can express statements like "is x less than 5 and not equal to y?" using [Boolean operators][boolean-operators]: `!` (not), `&&` (and), `||` (or).
+
+In Julia (and many other programming languages), `&&` has a [higher precedence][operator-precedence] than `||` (in the same way that `*` is applied before `+`).
+This means that `true || false && true` evaluates to `true` because it is parsed as `(true || false) && true`.
+It is common to include explicit brackets anyway so that the reader doesn't need to think about this.
```
Comment (BNAndras) [inline concepts/booleans/about.md:27]:
```suggestion
It is common to include explicit parentheses anyway so that the reader doesn't need to think about this.
```

File: concepts/booleans/about.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+julia> 2 < 1
+false
+
+julia> (5 * 5 == 25) || (2 < 1)
+true
+```
+
+### Short-circuit evaluation
+
+When `&&` or `||` are used to link Boolean expressions, Julia evaluates from left to right, and stops when it finds an unambiguous answer.
```
Comment (BNAndras) [inline concepts/booleans/about.md:85]:
```suggestion
When `&&` or `||` are used to link Boolean expressions, Julia evaluates from left to right, and stops when it finds an unambiguous answer.
```

We don't define what an unambiguous answer is before the example so it's a little hard to follow until the explanation. Should we just Julia always evaluates the left side and sometimes  the right side depending on the value of the left side and the operator used?

File: concepts/booleans/about.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+Consider this expression:
+
+```julia-repl
+julia> (5 > 12) && (1 != 0)
+true
+```
+
+Because `5 > 12` is `false`, clearly the combined expression must be `false`.
```
Comment (BNAndras) [inline concepts/booleans/about.md:94]: I don't believe we can say the combined expression is clearly `false` since we haven't introduced why the expression would be `false`. We also risk reinforcing whatever interpretation the student got from line 85 because they might think Julia stops when it hits a false value regardless of which operator is used so that's what "unambiguous" means.

File: concepts/booleans/about.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+The `1 != 0` expression is not relevant and is never evaluated.
```
Comment (BNAndras) [inline concepts/booleans/about.md:95]: `1 != 0` isn't evaluated when the left side evaluates to false. It is evaluated if the left side evaluates to true so we can't say it's never evaluated without qualifying that never.

File: concepts/booleans/introduction.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+```
+
+## Boolean logic
+
+Imagine we have the following Boolean expressions in Julia: `5 > x` and `x != 0`.
+If `x` was 3 they would both be `true`.
+We can express statements like "is x less than 5 and not equal to y?" using [Boolean operators][boolean-operators]: `!` (not), `&&` (and), `||` (or).
+
+In Julia (and many other programming languages), `&&` has a [higher precedence][operator-precedence] than `||` (in the same way that `*` is applied before `+`).
+This means that `true || false && true` evaluates to `true` because it is parsed as `(true || false) && true`.
```
Comment (BNAndras) [inline concepts/booleans/introduction.md:26]:
```suggestion
This means that `true || false && true` evaluates to `true` because it is parsed as `true || (false && true)`.
```

However, we haven't properly introduced the logical operators at this point so students who didn't follow the Boolean operators link will likely be confused how we evaluated that expression.

File: concepts/booleans/introduction.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+true
+
+julia> 1 != 0
+true
+
+julia> (5 > 3) && (1 != 0)
+true
+```
+
+Parentheses are optional and can make the code easier to read.
```
Comment (BNAndras) [inline concepts/booleans/introduction.md:66]: This should be included under the logical OR section. However, we've already mentioned this in the Boolean logic section so it can be dropped.

File: concepts/booleans/introduction.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+## Boolean logic
+
+Imagine we have the following Boolean expressions in Julia: `5 > x` and `x != 0`.
+If `x` was 3 they would both be `true`.
+We can express statements like "is x less than 5 and not equal to y?" using [Boolean operators][boolean-operators]: `!` (not), `&&` (and), `||` (or).
+
+In Julia (and many other programming languages), `&&` has a [higher precedence][operator-precedence] than `||` (in the same way that `*` is applied before `+`).
+This means that `true || false && true` evaluates to `true` because it is parsed as `(true || false) && true`.
+It is common to include explicit brackets anyway so that the reader doesn't need to think about this.
```
Comment (BNAndras) [inline concepts/booleans/introduction.md:27]:
```suggestion
It is common to include explicit parentheses anyway so that the reader doesn't need to think about this.
```

**Comment by colinleach** [comment]
I need to think about this and try again.

**Comment by colinleach** [comment]
Your key point is that comparison operators aren't taught till the next concept. I'm a bit less worried about this in `about.md`, which is allowed to forward-reference, but these should definitely be removed from `introduction.md`.

One take-home message: *don't* blindly reuse text from the older exercise drafts, as I did here.. There are reasons why this syllabus hasn't launched.

**Comment by colinleach** [comment]
I did a major rewrite (essentially from scratch) of both Markdown files, in response to the all-too-valid criticisms from @BNAndras. 

Further comments welcome, from anyone.

File: concepts/booleans/about.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+julia> typeof(true)
+Bool
+```
+
+In contrast to several other languages, Julia deliberately has no concept of "truthiness".
+Only expressions which evaluate to `true` or `false` will be treated as a `Bool`.
+
+Specifically, empty arrays or strings will *not* be interpreted as `false`.
+There must be an appropriate test such as `isempty()` if you want special handlind for empty values.
```
Comment (BNAndras) [inline concepts/booleans/about.md:23]:
```suggestion
There must be an appropriate test such as `isempty()` if you want special handling for empty values.
```

File: concepts/booleans/about.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+```julia-repl
+julia> !true
+false
+
+julia> !false
+true
+```
+
+These operators will be familiar to users of many other languages, though *not* Python.
```
Comment (BNAndras) [inline concepts/booleans/about.md:62]: This sentence is confusing. `and` and `or` exist in other languages. Perhaps we say the operators and these concepts should be familiar coming from other languages but they might have different names there.

**Review by BNAndras — COMMENTED** [review]
Not at a computer but these jumped out at me.

**Comment by colinleach** [comment]
Aargh! I have done something very, very wrong with this, and the files changed now include GH workflows and all sorts of nonsense.

Unless I can dig myself out of this soon, I may close this PR and open a new one. It was just about ready to review, before I broke it!

**Comment by colinleach** [comment]
Closing now, I wrecked this PR. Continues with #831.
